File: 2. Sorting Techniques/Sorting II/C++/3. Recursive Insertion Sort.cpp
/*
Recursive Insertion Sort
*/

#include <bits/stdc++.h>
using namespace std;

void swap(int arr[],int i,int j){
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

//Function to sort an array using recursive insertion sort algorithm.
void recursiveInsertionSort(int arr[], int i, int n)
{
    if(i==n){
        return;
    }
    int j = i;
    while(j>0 && arr[j-1]>arr[j]){
        swap(arr,j,j-1);
        j--;
    }
  
    recursiveInsertionSort(arr,i+1,n);
}

int main() {
	// your code goes here
    int n;
    cin >> n;
    int arr[n];
    for(int i=0;i<n;i++){
        cin >> arr[i];
    }
    recursiveInsertionSort(arr,1,n);
    for(int i=0;i<n;i++){
        cout << arr[i] << " "; 
    }
	return 0;
}

File: 2. Sorting Techniques/Sorting II/Python/3. Recursive Insertion Sort.py
# cook your dish here
#Recursive Insertion Sort


def recursiveInsertionSort(arr,i,n):
    if(i==n):
        return
    j = i
    while(j>0 and arr[j-1]>arr[j]):
        (arr[j],arr[j-1]) = (arr[j-1],arr[j])
        j = j -1 

    recursiveInsertionSort(arr,i+1,n)

n = int(input())
arr = list(map(int,input().strip().split()))
recursiveInsertionSort(arr,1,n)
for num in arr:
    print(num,end=" ")